Fix prelu ref when a.ndim < 2

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89809

Fixes https://github.com/pytorch/pytorch/issues/89560

Previously the test case for "input is 1-D or scalar + weight is not scalar" did not exist; adding it introduced some failures:
- forward AD (fixed in this PR)
- vmap (filed https://github.com/pytorch/pytorch/issues/89895)
- ref/meta (fixed this PR, though this also regresses nvFuser support)

soulitzer committed:
Fix prelu ref when a.ndim < 2

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge -g

soulitzer committed:
Update on "Fix prelu ref when a.ndim < 2"


Fixes https://github.com/pytorch/pytorch/issues/89560

[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -453,8 +453,9 @@ def sample_inputs_prelu(op_info, device, dtype, requires_grad, **kwargs):
 
-        channel_size = shape[1] if len(shape) >= 2 else 1
-        yield SampleInput(make_arg(shape), args=(make_arg((channel_size,)),))
+        if len(shape) >= 1:
```
Comment here for what's happening

mruberry commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -960,7 +960,12 @@ def prelu(a: TensorLikeType, weight: TensorLikeType) -> TensorLikeType:
     )
     if a.ndim == 0:
-        weight = weight[0] if weight.ndim == 1 else weight
+        # NB: This diverges from the current prelu which actually allows this case
```
Would you elaborate on what's happening in this case and what the issue with nvFuser is?

ngimel replied:
We should skip nvfuser test, or nvfuser can do something in their integration, but we shouldn't change general refs to deviate from eager semantics.

soulitzer replied:
The issue with nvfuser was that I'm trying to call `select`, `view_as`, `view`, etc, to get weight to be a scalar, but those aren't supported.

ngimel replied:
But that happens only when a.ndim == 0 and weight.ndim is not zero? It's fine for nvfuser to not support this case for now (they should fix it).

soulitzer committed:
Update on "Fix prelu ref when a.ndim < 2"


Fixes https://github.com/pytorch/pytorch/issues/89560

Previously the test case for "input is 1-D or scalar + weight is not scalar" did not exist; adding it introduced some failures:
- forward AD (fixed in this PR)
- vmap (filed https://github.com/pytorch/pytorch/issues/89895)
- ref/meta (fixed this PR, though this also regresses nvFuser support)

[ghstack-poisoned]

soulitzer committed:
Update on "Fix prelu ref when a.ndim < 2"


Fixes https://github.com/pytorch/pytorch/issues/89560

Previously the test case for "input is 1-D or scalar + weight is not scalar" did not exist; adding it introduced some failures:
- forward AD (fixed in this PR)
- vmap (filed https://github.com/pytorch/pytorch/issues/89895)
- ref/meta (fixed this PR, though this also regresses nvFuser support)

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge -f "Unrelated failures"

soulitzer committed:
Update on "Fix prelu ref when a.ndim < 2"


Fixes https://github.com/pytorch/pytorch/issues/89560

Previously the test case for "input is 1-D or scalar + weight is not scalar" did not exist; adding it introduced some failures:
- forward AD (fixed in this PR)
- vmap (filed https://github.com/pytorch/pytorch/issues/89895)
- ref/meta (fixed this PR, though this also regresses nvFuser support)

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge -g

soulitzer committed:
Update on "Fix prelu ref when a.ndim < 2"


Fixes https://github.com/pytorch/pytorch/issues/89560

Previously the test case for "input is 1-D or scalar + weight is not scalar" did not exist; adding it introduced some failures:
- forward AD (fixed in this PR)
- vmap (filed https://github.com/pytorch/pytorch/issues/89895)
- ref/meta (fixed this PR, though this also regresses nvFuser support)

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge -g

soulitzer committed:
Update on "Fix prelu ref when a.ndim < 2"


Fixes https://github.com/pytorch/pytorch/issues/89560

Previously the test case for "input is 1-D or scalar + weight is not scalar" did not exist; adding it introduced some failures:
- forward AD (fixed in this PR)
- vmap (filed https://github.com/pytorch/pytorch/issues/89895)
- ref/meta (fixed this PR, though this also regresses nvFuser support)

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge -g

soulitzer commented:
@pytorchbot merge -f "Unrelated failure: rocm timing out"
